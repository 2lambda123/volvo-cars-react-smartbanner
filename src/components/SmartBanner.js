import React, { Component, PropTypes } from 'react';
import ua from 'ua-parser-js';
import cookie from 'cookie-cutter';

import '../styles/style.scss';

class SmartBanner extends Component {
  constructor(props) {
    super(props);

    this.setType(this.props.force);
  }

  static propTypes = {
    daysHidden: PropTypes.number,
    daysReminder: PropTypes.number,
    appStoreLanguage: PropTypes.string,
    button: PropTypes.string,
    storeText: PropTypes.objectOf(PropTypes.string),
    price: PropTypes.objectOf(PropTypes.string),
    force: PropTypes.string,
    title: PropTypes.string,
    author: PropTypes.string
  }

  static defaultProps = {
    daysHidden: 15,
    daysReminder: 90,
    appStoreLanguage: navigator.language.slice(-2) || navigator.userLanguage.slice(-2) || 'us',
    button: 'View',
    storeText: {
      ios: 'On the App Store',
      android: 'In Google Play',
      windows: 'In Windows Store'
    },
    price: {
      ios: 'Free',
      android: 'Free',
      windows: 'Free'
    },
    force: '',
    title: '',
    author: ''
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.force !== this.props.force) {
      this.setType(nextProps.force);
    }
  }

  type = '';
  appId = '';
  settings = {};

  parseAppId() {
    let meta = document.querySelector('meta[name="' + this.settings.appMeta + '"]');
    if (!meta) {
      return '';
    }

    if (this.type === 'windows') {
      this.appId = meta.getAttribute('content');
    } else {
      this.appId = /app-id=([^\s,]+)/.exec(meta.getAttribute('content'))[1];
    }

    return this.appId;
  }

  setType(deviceType) {
    let mixins = {
      ios: {
        appMeta: 'apple-itunes-app',
        iconRels: ['apple-touch-icon-precomposed', 'apple-touch-icon'],
        getStoreLink: () => {
          return 'https://itunes.apple.com/' + this.props.appStoreLanguage + '/app/id' + this.appId;
        }
      },
      android: {
        appMeta: 'google-play-app',
        iconRels: ['android-touch-icon', 'apple-touch-icon-precomposed', 'apple-touch-icon'],
        getStoreLink: () => {
          return 'http://play.google.com/store/apps/details?id=' + this.appId;
        }
      },
      windows: {
        appMeta: 'msApplication-ID',
        iconRels: ['windows-touch-icon', 'apple-touch-icon-precomposed', 'apple-touch-icon'],
        getStoreLink: () => {
          return 'http://www.windowsphone.com/s?appid=' + this.appId;
        }
      }
    };

    let agent = ua(navigator.userAgent);

    if (deviceType.length) {
      this.type = deviceType;
    } else if (agent.os.name === 'Windows Phone' || agent.os.name === 'Windows Mobile') {
      this.type = 'windows';
    //iOS >= 6 has native support for Smart Banner
    } else if (agent.os.name === 'iOS' && parseInt(agent.os.version, 10) < 6) {
      this.type = 'ios';
    } else if (agent.os.name === 'Android') {
      this.type = 'android';
    }

    this.settings = mixins[this.type];
  }

  hide() {
    document.querySelector('html').classList.remove('smartbanner-show');
  }

  show() {
    document.querySelector('html').classList.add('smartbanner-show');
  }

  close() {
    this.hide();
    cookie.set('smartbanner-closed', 'true', {
      path: '/',
      expires: +new Date() + this.props.daysHidden * 1000 * 60 * 60 * 24
    });
  }

  install() {
    this.hide();
    cookie.set('smartbanner-installed', 'true', {
      path: '/',
      expires: +new Date() + this.props.daysReminder * 1000 * 60 * 60 * 24
    });
  }

  render() {
    // Don't show banner if device isn't iOS or Android, website is loaded in app,
    // user dismissed banner, or we have no app id in meta
    if (!this.type
      || navigator.standalone
      || cookie.get('smartbanner-closed')
      || cookie.get('smartbanner-installed')) {
      return false;
    }

    if (this.parseAppId() === '') {
      return false;
    }

    this.show();

    let link = this.settings.getStoreLink();
    let inStore = this.props.price[this.type] + ' - ' + this.props.storeText[this.type];
    let icon;
    for (let i = 0, max = this.settings.iconRels.length; i < max; i++) {
      let rel = document.querySelector('link[rel="' + this.settings.iconRels[i] + '"]');
      if (rel) {
        icon = rel.getAttribute('href');
        break;
      }
    }

    let wrapperClassName = 'smartbanner smartbanner-' + this.type;
    let iconStyle = {
      backgroundImage: 'url(' + icon + ')'
    };

    return (
      <div className={wrapperClassName}>
        <div className="smartbanner-container">
          <a className="smartbanner-close" onClick={::this.close}>&times;</a>
          <span className="smartbanner-icon" style={iconStyle}></span>
          <div className="smartbanner-info">
            <div className="smartbanner-title">{this.props.title}</div>
            <div>{this.props.author}</div>
            <span>{inStore}</span>
          </div>

          <a href={link} onClick={::this.install} className="smartbanner-button">
            <span className="smartbanner-button-text">{this.props.button}</span>
          </a>
        </div>
      </div>
    );
  }
}

export default SmartBanner;
