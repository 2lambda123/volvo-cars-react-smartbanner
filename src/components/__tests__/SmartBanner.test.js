/* eslint-disable import/first, react/no-find-dom-node, no-eval */
jest.mock('cookie-cutter', () => {
  return {
    set: jest.fn(),
    get: jest.fn(),
  };
});

import React from 'react';
import { mount } from 'enzyme';
import SmartBanner from '../SmartBanner';

describe('SmartBanner', function () { // eslint-disable-line func-names
  let cookie;

  beforeEach(() => {
    jest.resetModules();

    cookie = require('cookie-cutter'); // eslint-disable-line global-require

    document.querySelector('head').innerHTML = `
      <meta name="apple-itunes-app" content="app-id=284882215">
      <meta name="google-play-app" content="app-id=com.facebook.katana">
      <meta name="kindle-fire-app" content="app-id=B0094BB4TW">
      <meta name="msApplication-ID" content="82a23635-5bd9-df11-a844-00237de2db9e">
      <meta name="msApplication-PackageFamilyName" content="facebook_9wzdncrfhv5g">`;
    document.body.innerHTML = '<div id="root"></div>';

    this.params = {
      force: '',
      title: '',
      author: '',
      url: '',
    };

    this.makeSubject = (_props = {}) => {
      const props = {
        ...this.params,
        ..._props,
      };

      return mount(
        <SmartBanner
          { ...props }
        />,
        {
          attachTo: document.getElementById('root'),
        },
      );
    };
  });

  it('should be rendered', () => {
    const subject = this.makeSubject();

    expect(subject.length).toBeTruthy();
  });

  describe('snapshots', () => {
    it('should be matched the snapshot (no type)', () => {
      const subject = this.makeSubject();

      expect(subject.html()).toMatchSnapshot();
    });

    it('should be matched the snapshot (android)', () => {
      const subject = this.makeSubject({
        force: 'android',
      });

      expect(subject.state('type')).toBe('android');
      expect(subject.html()).toMatchSnapshot();
    });

    it('should be matched the snapshot (ios)', () => {
      const subject = this.makeSubject({
        force: 'ios',
      });

      expect(subject.state('type')).toBe('ios');
      expect(subject.html()).toMatchSnapshot();
    });

    it('should be matched the snapshot (kindle)', () => {
      const subject = this.makeSubject({
        force: 'kindle',
      });

      expect(subject.state('type')).toBe('kindle');
      expect(subject.html()).toMatchSnapshot();
    });

    it('should be matched the snapshot (windows)', () => {
      const subject = this.makeSubject({
        force: 'windows',
      });

      expect(subject.state('type')).toBe('windows');
      expect(subject.html()).toMatchSnapshot();
    });
  });

  describe('close smartbanner', () => {
    it('should change html classList and set cookie after click the close button', () => {
      const subject = this.makeSubject({
        force: 'android',
      });

      subject.find('.smartbanner-close').simulate('click');

      expect(window.document.querySelector('html').classList).not.toContain('smartbanner-show');
      expect(cookie.set).toBeCalledWith('smartbanner-closed', 'true', {
        path: '/',
        expires: 1495756800000,
      });
    });
  });

  describe('click install on smartbanner', () => {
    it('should change html classList and set cookie after click the close button', () => {
      const subject = this.makeSubject({
        force: 'android',
      });

      subject.find('.smartbanner-button').simulate('click');

      expect(window.document.querySelector('html').classList).not.toContain('smartbanner-show');
      expect(cookie.set).toBeCalledWith('smartbanner-installed', 'true', {
        path: '/',
        expires: 1502236800000,
      });
    });
  });
});
