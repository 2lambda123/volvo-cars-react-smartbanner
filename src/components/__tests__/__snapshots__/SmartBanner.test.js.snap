exports[`SmartBanner snapshots should be matched the snapshot (android) 1`] = `
"<div class=\"smartbanner smartbanner-android\"><div class=\"smartbanner-container\"><a class=\"smartbanner-close\">×</a><span class=\"smartbanner-icon\" style=\"background-image: url(undefined);\"></span><div class=\"smartbanner-info\"><div class=\"smartbanner-title\"></div><div class=\"smartbanner-author\"></div><div class=\"smartbanner-description\">
      Free - In Google Play</div></div><div class=\"smartbanner-wrapper\"><a href=\"http://play.google.com/store/apps/details?id=com.facebook.katana\" class=\"smartbanner-button\"><span class=\"smartbanner-button-text\">View</span></a></div></div></div>"
`;

exports[`SmartBanner snapshots should be matched the snapshot (ios) 1`] = `
"<div class=\"smartbanner smartbanner-ios\"><div class=\"smartbanner-container\"><a class=\"smartbanner-close\">×</a><span class=\"smartbanner-icon\" style=\"background-image: url(undefined);\"></span><div class=\"smartbanner-info\"><div class=\"smartbanner-title\"></div><div class=\"smartbanner-author\"></div><div class=\"smartbanner-description\">
      Free - On the App Store</div></div><div class=\"smartbanner-wrapper\"><a href=\"https://itunes.apple.com/US/app/id284882215\" class=\"smartbanner-button\"><span class=\"smartbanner-button-text\">View</span></a></div></div></div>"
`;

exports[`SmartBanner snapshots should be matched the snapshot (kindle) 1`] = `
"<div class=\"smartbanner smartbanner-kindle\"><div class=\"smartbanner-container\"><a class=\"smartbanner-close\">×</a><span class=\"smartbanner-icon\" style=\"background-image: url(undefined);\"></span><div class=\"smartbanner-info\"><div class=\"smartbanner-title\"></div><div class=\"smartbanner-author\"></div><div class=\"smartbanner-description\">
      Free - In the Amazon Appstore</div></div><div class=\"smartbanner-wrapper\"><a href=\"amzn://apps/android?asin=B0094BB4TW\" class=\"smartbanner-button\"><span class=\"smartbanner-button-text\">View</span></a></div></div></div>"
`;

exports[`SmartBanner snapshots should be matched the snapshot (no type) 1`] = `"<div></div>"`;

exports[`SmartBanner snapshots should be matched the snapshot (windows) 1`] = `
"<div class=\"smartbanner smartbanner-windows\"><div class=\"smartbanner-container\"><a class=\"smartbanner-close\">×</a><span class=\"smartbanner-icon\" style=\"background-image: url(undefined);\"></span><div class=\"smartbanner-info\"><div class=\"smartbanner-title\"></div><div class=\"smartbanner-author\"></div><div class=\"smartbanner-description\">
      Free - In Windows Store</div></div><div class=\"smartbanner-wrapper\"><a href=\"http://www.windowsphone.com/s?appid=82a23635-5bd9-df11-a844-00237de2db9e\" class=\"smartbanner-button\"><span class=\"smartbanner-button-text\">View</span></a></div></div></div>"
`;
